File: 0219-contains-duplicate-ii/0219-contains-duplicate-ii.py
from collections import defaultdict


class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        indices = defaultdict(list)

        for index, number in enumerate(nums):
            indices[number].append(index)

        for i in indices.values():
            i.sort()
            if len(i) > 1:
                for j in range(len(i) - 1):
                    if i[j + 1] - i[j] <= k: return True

        return False

File: 0219-contains-duplicate-ii/0219-contains-duplicate-ii.cpp
class Solution {
public:
    bool containsNearbyDuplicate(vector<int>& nums, int k) {
        unordered_map<int, int> indices;

        for (int i = 0; i < nums.size(); i++) {
            if (indices.find(nums[i]) != indices.end()) {
                if (i - indices[nums[i]] <= k) return true;
            }
            indices[nums[i]] = i;
        }

        return false;
    }
};